### DISCUSSION TITLE
"Open Role" integration with the Runtime

### DISCUSSION THREAD

**Discussion by paluh** [marlowe-lang/marlowe-cardano#711] [Architecture]
#### Core Feature

* We have to add "Open Role" validator to the script registry.

* During transaction creation we have to create an UTxO with role token and min ADA. Inline datum should contain the name of the "thread token" - we can use by default `""`.

* During the transaction creation we should output thread token to the Marlowe UTxO. We could do this with `IDeposit` and initial input application or we can add this token to the account system manually like we do when we handle min ADA deposit. We should deposit this token to the tx creator account.

* We should index these UTxOs by role token so we can quickly find them when we miss the token from the `WalletContext`.

* When we detect that the role token should be released we provide this input + reference input to the "Open Role" validator to the transaction.

#### User facing API changes:

* We should change the REST API schema for the `/contracts/create` request so the user can either specify a destination address or indicate usage of open role validator.

* We should add similar change to the `marlowe-runtime-cli`.

**Comment by bwbush** [comment]
> we can use by default ""

We'll need a safety check that the role token name isn't already used in the contract.

**Reply by paluh** [reply]
That is true *but* we want to probably use different policy for the thread token at end - it is discussed here: https://github.com/input-output-hk/marlowe-cardano/discussions/711

**Comment by bwbush** [comment]
>We should index these UTxOs by role token so we can quickly find them when we miss the token from the WalletContext.

Is this strictly necessary, or just nice to have? Adding an index might be quite a bit of extra work, especially at this stage where open roles are somewhat experimental.

**Reply by paluh** [reply]
I mean we can reuse indexes which we have if they allow us to query by script address and datum. We can leave this `txOutRef` definition to the user - this kind of query could be covered by `blockfrost` API for asset history for example.
Of course it won't be easy to provide support for this in `ts-sdk` without dependency on blockfrost API key in such a case. @nhenin, @hrajchert - what do you think Guys about this?

**Comment by bwbush** [comment]
>We should add similar change to the marlowe-runtime-cli.

I suggest we omit the CLI from the MVP.

**Comment by bwbush** [comment]
>We should change the REST API schema for the /contracts/create request so the user can either specify a destination address or indicate usage of open role validator.

I'm thinking of forward and future backward compatibility.
- Should this be included in `/contracts/create`? I'm wondering if we should avoid making the creation endpoint ever more complex as we add new features such as open roles.
- How experimental is this? I'm wondering if we should have an `/experimental` path in the API.

**Reply by paluh** [reply]
@jhbertra what is your opinion on that?
